src/bt.rs

```diff
@@ -91,310 +97,290 @@ impl<V> Node<V> {
+    /// along the path starting from the root node.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
+    /// This function traverses the tree from the root until reaching the
     /// node at the end of the path. If the node is unreachable or already
-    /// contains a value, an error wrapping the value is returned. Otherwise,
-    /// a new node containing the value is inserted.
+    /// contains a value, the function returns an error wrapping the value;
+    /// otherwise, it stores the value in a new node and returns a reference
+    /// to the node.
+    ///
+    /// # Returns
+    /// - `Ok(NodeRef)` a reference is returned when the node is inserted
+    /// at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
+    /// value at the end of the path.
+    /// - `Err(UnreachableNode(value))` when the end of the path is unreachable.
+    ///
+    /// In the error cases, no insertion occurs and the value is returned to
+    /// the caller of this function.
+    pub fn insert(&mut self, path: &Path, value: V) -> Result<NodeRef, BinaryTreeError<V>> {
+        self.insert_at(NodeRef::ROOT, path, value)
+    }
+
+    /// Inserts the value at the location reached by traversing the tree
+    /// along the path starting from the specified node reference.
+    ///
+    /// This function traverses the tree from the node specified until
+    /// reaching the node at the end of the path. If the node is unreachable
+    /// or already contains a value, the function returns an error wrapping
+    /// the value; otherwise, it stores the value in a new node and returns
+    /// a reference to the node.
     ///
     /// # Returns
-    /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the subtree already contains a
+    /// - `Ok(NodeRef)` a reference is returned when the node is inserted
+    /// at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
     /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
+    /// - `Err(UnreachableNode(value))` when the end of the path is unreachable.
+    /// - `Err(InvalidNodeRef(value))` when the node reference is invalid.
     ///
     /// In the error cases, no insertion occurs and the value is returned to
     /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        enum Ref<'a, V> {
-            This(&'a mut Node<V>),
-            Other(&'a mut Option<Box<Node<V>>>),
+    pub fn insert_at(
+        &mut self,
+        node_ref: NodeRef,
+        path: &Path,
+        value: V,
+    ) -> Result<NodeRef, BinaryTreeError<V>> {
+        if !(self.is_valid_node_ref(node_ref)
+            || (self.nodes.is_empty() && node_ref == NodeRef::ROOT))
```

Comment:
nit: I wonder if the second term in this `||` should be implied by the first? I.e., if the list of nodes is empty, and the refernece is to the root, then it's a valid reference, correct?

Put simply: consider moving `(self.nodes.is_empty() && node_ref == NodeRef::ROOT)` into `is_valid_node_ref()`.